### PR TITLE
added feature of terminating when the link is deleted issue number #352 issue has been solved 

### PR DESCRIPTION
**Pull request by chandankumarm55**
Hey, thanks for assigning this issue to me!

I successfully implemented this feature. Most of the files show changes due to the addition of a cursor at the end. In the frontend, we used a useEffect hook to validate whether the link is valid or not. We also applied a throttling concept to check this condition every second. You can modify the interval to your preference, but 1 second is optimal for performance. 

Things you need to change for production mode:

You need to update the URL used for sending the request in the file client/src/pages/messaging/index.tsx at line 270. For production, the URL should be https://chat-e2ee-2.azurewebsites.net/api/chat-link/status/${channelID}. Make sure you use backticks for this. For now, I’ve used localhost for running it locally.

Screenshot 2024-09-13 125631

Please refer to the commits for a better understanding. You can also use an .env file and access the URL via process.env.BACKEND_URL, but for now, I’ve added the URL directly to reduce confusion.

I have also tested the app locally. Only one test failed due to the absence of an absolute link, which we declared as undefined. This is causing the following error:

The detailed failed test case can be found in the file backend/api/chatLink/utils/link.test.ts at line 19. The error is caused by absoluteLink: undefined, which is why the test failed.

Screenshot 2024-09-13 130357

Screenshot 2024-09-13 130424

i think this is not matter that much . and console when we start the app looks like 

Screenshot 2024-09-13 134939

I implemented the feature where the throttling function runs every second and renders the component when linkActive becomes invalid. After 2 seconds, it will navigate to the / route (home route). You can adjust this timing based on your preference.

I chose 2 seconds as an optimal delay to inform the user that the link is invalid. Using a longer delay might send more requests to the backend, leading to performance issues. On the other hand, if we use a 1-second delay, there might not be enough time for the user to read the message. Therefore, I think 2 seconds strikes a good balance, but you can modify it according to your needs.

https://github.com/user-attachments/assets/ed78425e-e80e-48df-b468-f195c0c07c7d

now Related to delete link feature #352 issue has been solved 

fell free to as anything , to ask regarding this issue impentation . Thank you for guiding at every step @muke1908

### CODE REVIEW

**Comment by chandankumarm55** [comment]
@muke1908  i have installed many libraries i faced difficulty in running the application at starting so it showing there is conflicts in package.json you can omit or remove those unwanted later please merge it as soon as possible .

File: backend/api/chatLink/utils/link.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { v4 as uuidv4 } from 'uuid';
 import { generatePIN } from './pin';
 
-const { CHAT_LINK_DOMAIN } = process.env;
+const CHAT_LINK_DOMAIN = process.env.CHAT_LINK_DOMAIN || 'http://localhost:3000';
```
Comment (muke1908) [inline backend/api/chatLink/utils/link.ts:4]: Please remove this fallback.

File: backend/db/index.ts [hunk 2]
```diff
@@ -4,10 +4,17 @@ import {
     findOneFromDB as _findOneFromDB, insertInDb as _insertInDb, updateOneFromDb as _updateOneFromDb
 } from './inMemDB';
 
-const uri = process.env.MONGO_URI;
-const dbName = process.env.MONGO_DB_NAME;
+// added local db address if you didnt create a .env file
+const uri = process.env.MONGO_URI || 'mongodb://localhost:27017';
+const dbName = process.env.MONGO_DB_NAME || 'realtimechatting';
```
Comment (muke1908) [inline backend/db/index.ts:9]: Please remove the arbitrary defaults. Let it be set explicitly in .env

File: backend/db/index.ts [hunk 3]
```diff
@@ -4,10 +4,17 @@ import {
...
+
+console.log('uri', uri);
```
Comment (muke1908) [inline backend/db/index.ts:11]: unnecessary `console.log`

File: client/src/pages/messaging/index.tsx [hunk 4]
```diff
@@ -258,6 +260,69 @@ const Chat = () => {
...
+  }, 1000); // Check every 1 seconds (adjust this interval as needed)
+
+  return () => {
+    clearInterval(intervalId); // Cleanup the interval on component unmount
+  };
+}, [channelID, history]);
+
+
+// Helper function to throttle a function call
+function throttle<T extends (...args: any[]) => void>(func: T, limit: number): (...args: Parameters<T>) => void {
```
Comment (muke1908) [inline client/src/pages/messaging/index.tsx:313]: can be moved to ./utils.

File: client/src/pages/messaging/index.tsx [hunk 5]
```diff
@@ -258,6 +260,69 @@ const Chat = () => {
...
     };
   });
+
+// added a useeffect to check whather the 
+useEffect(() => {
+  const validateLink = async () => {
+    try {
+      // FOR PRODUCTION MODE URL SHOULD BE https://chat-e2ee-2.azurewebsites.net/api/chat-link/status/${channelID}
+      // FOR RUNING LOCALLY URL SHOULD BE http://localhost:3001/api/chat-link/status/${channelID}
+      const URL = `http://localhost:3001/api/chat-link/status/${channelID}`
```
Comment (muke1908) [inline client/src/pages/messaging/index.tsx:270]: urls should not be hard-coded. Also UI layer  does not need to know anything about the server. That's why we have the service. 
You can expose another method in `service`, and use that here.

**Review by muke1908 — CHANGES_REQUESTED** [review]

**Comment by chandankumarm55** [comment]
@muke1908  can please guide me what and all need to make change in order to create new method ?

**Comment by muke1908** [comment]
> @muke1908 can please guide me what and all need to make change in order to create new method ?

I think you already did it earlier here https://github.com/muke1908/chat-e2ee/issues/352#issuecomment-2336803634

**Comment by chandankumarm55** [comment]
yes i did but  the error not found so first i created method in sdk.ts and added that in types.ts after that  run build command then also it didnt reflect on my frontend

**Comment by muke1908** [comment]
> yes i did but the error not found so first i created method in sdk.ts and added that in types.ts after that run build command then also it didnt reflect on my frontend

Did you try this https://github.com/muke1908/chat-e2ee/issues/352#issuecomment-2345574879

**Comment by chandankumarm55** [comment]
@muke1908  after trying only i purposed this method . i tried for more then 2 days but  it showing same error method is not found . please mention a detailed description how to create new method in service folder and how backend folder interact with service folder please

**Comment by muke1908** [comment]
> @muke1908 after trying only i purposed this method . i tried for more then 2 days but it showing same error method is not found . please mention a detailed description how to create new method in service folder and how backend folder interact with service folder please


1. go to `./service`, run `npm link`
2. go to `./client`, delete `node_modules`, run `npm link @chat-e2ee/service`, run `npm install`

Once it's done, if you make change in `./service` and run build it will immediately reflect in `./client`

**Comment by muke1908** [comment]
closing as redundant

**Pull request closed without merging**